### PR TITLE
purescript: add aarch64-darwin to platforms list

### PR DESCRIPTION
**Pull request by lucperkins**
###### Description of changes

The `purescript` package doesn't currently list `aarch64-darwin` amongst the available platforms, but it builds and runs just fine on my M1.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by drupol — APPROVED** [review]
LGTM

**Comment by justinwoo** [comment]
seems okay to me, but does it require any extra setup that one should know about?

**Review by cdepillabout — CHANGES_REQUESTED** [review]
I believe this came up on a previous purescript PR, but my understanding of the situation is:

- In nixpkgs, the top-level `purescript` derivation is not built from source, but we just `patchelf` binaries downloaded from github.
- The only binaries that are provided by the upstream purescript releases are for x86_64-linux and x86_64-darwin.
- The aarch64-darwin machines have some sort of internal emulation layer that allows them to run x86_64-darwin binaries as-is.

The question is what does `platforms` mean semantically in nixpkgs?  Does it represent the platforms of the binaries produced by the derivation?  Or does it represent the platforms that the binaries can run on?

I believe the decision was that it represents the platforms of the binaries produced by the derivation (so the end-user is responsible for setting up their system so that nix knows to allow both `x86_64-darwin` binaries and `aarch64-darwin` binaries).

I'm not very familiar with Darwin, so if you don't think this is how it should be, feel free to ping @-NixOS/darwin to weigh-in on this.

**Comment by lucperkins** [comment]
@justinwoo Not that I'm aware of. Plain old `nix-build -A purescript` works fine for me on `aarch64-darwin`.

@cdepillabout That's a great question about what "platform support" ultimately means. Personally, I'm agnostic on that and would be happy to close this if platform support means something different from how I interpret it. For me, it's basically synonymous with "I can build and then run this on my machine." `nix-build -A purescript && ./result/bin/purs` works on my Mac M1, which satisfies my personal criterion. But again, 100% happy to defer to others who are far closer to these issues than I am as a bystander!

**Comment by cdepillabout** [comment]
@lucperkins Feel free to ping @-NixOS/darwin to make a call on this.

**Comment by lucperkins** [comment]
@NixOS/darwin-maintainers Would love some input here. TL;DR: the `purescript` package builds/runs fine on Mac M1, but it's unclear that `aarch64-darwin` should be added to the `platforms`.

**Comment by siraben** [comment]
Result of `nixpkgs-review pr 198827` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>purescript</li>
  </ul>
</details>

**Comment by reckenrode** [comment]
If it’s not actually providing an aarch64-darwin binary, I don’t think the derivation should advertise support for that platform. One can pass `--system x86_64-darwin` to Nix to run the x86_64-darwin version instead. There’s also #161657, which will provide x86_64-darwin packages for use with Rosetta 2 on aarch64-darwin.

**Review by siraben — APPROVED** [review]
IMO this is fine. Even if running through Rosetta the point is that the package works correctly on the platform, and this can allow for other packages to now depend on purescript on aarch64-darwin.

**Comment by siraben** [comment]
purescript is a Haskell package, shouldn't we try to build it from source as well?

**Comment by toonn** [comment]
I'm on the "platforms lists what something is built for" side of the fence. Adding `aarch64-darwin` to most of the darwin packages in Nixpkgs is going to make it harder to figure out whether a package builds natively for aarch64-darwin. And AFAIU it's not actually difficult to use `x86_64-darwin` packages on newer systems, I assume you can add the x86 platform in your configuration somehow, like how Linux x86_64 systems can run i686 packages too?

Re building Purescript from source, I believe the problem is upstream sticks with very old GHC releases for longer than the Haskell maintainers are willing to put up with. So it's not actually an option without major porting effort, if that would even be accepted upstream regularly.

**Comment by prusnak** [comment]
I think this should be closed in favor of https://github.com/NixOS/nixpkgs/pull/161657

**Review by uri-canva — CHANGES_REQUESTED** [review]
Yeah I'm against it as well for the same reasons. We can add it once upstream publishes `aarch64-darwin` binary packages.

**Comment by marsam** [comment]
previous discussion https://github.com/NixOS/nixpkgs/issues/183481

**Comment by lucperkins** [comment]
General consensus seems to be to close this. The reasoning 100% makes sense to me 👍🏼

**Pull request closed without merging**
